pkg/dyninst/decode/types.go

```diff
@@ -0,0 +1,955 @@
+	switch s := irType.(type) {
+	case *ir.GoSwissMapHeaderType:
+		dirPtrField, err := getFieldByName(s.RawFields, "dirPtr")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		dirPtrOffset := dirPtrField.Offset
+		dirPtrSize := dirPtrField.Type.GetByteSize()
+		dirLenField, err := getFieldByName(s.RawFields, "dirLen")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		dirLenOffset := dirLenField.Offset
+		dirLenSize := dirLenField.Type.GetByteSize()
+
+		usedField, err := getFieldByName(s.RawFields, "used")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		usedOffset := usedField.Offset
+		usedSize := usedField.Type.GetByteSize()
+
+		slotsField, err := getFieldByName(s.GroupType.RawFields, "slots")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		slotsFieldType, ok := types[slotsField.Type.GetID()]
+		if !ok {
+			return nil, errors.New("type map slot field not found in types: " + s.GroupType.Name)
+		}
+		ctrlField, err := getFieldByName(s.GroupType.RawFields, "ctrl")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		ctrlOffset := ctrlField.Offset
+		ctrlSize := ctrlField.Type.GetByteSize()
+		entryArray, ok := slotsFieldType.(*ir.ArrayType)
+		if !ok {
+			return nil, errors.New("type map slot field is not an array type: " + slotsFieldType.GetName())
+		}
+		noalgstructType, ok := entryArray.Element.(*ir.StructureType)
+		if !ok {
+			return nil, errors.New("type map entry array element is not a structure type: " + entryArray.Element.GetName())
+		}
+		keyField, err := getFieldByName(noalgstructType.RawFields, "key")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		if keyField == nil {
+			return nil, errors.New("type map entry array element has no key field: " + entryArray.Element.GetName())
+		}
+		elem, err := getFieldByName(noalgstructType.RawFields, "elem")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+
+		tableType := s.TablePtrSliceType.Element.(*ir.PointerType)
+		tableStructType := tableType.Pointee.(*ir.StructureType)
+		groupField, err := getFieldByName(tableStructType.RawFields, "groups")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		groupType, ok := groupField.Type.(*ir.GoSwissMapGroupsType)
+		if !ok {
+			return nil, fmt.Errorf("group field type is not a swiss map groups type: %s", groupField.Type.GetName())
+		}
+		groupFieldOffset := groupField.Offset
+		groupFieldSize := groupField.Type.GetByteSize()
+		dataField, err := getFieldByName(groupType.RawFields, "data")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		dataFieldOffset := dataField.Offset
+		dataFieldSize := dataField.Type.GetByteSize()
```

Comment:
you have unchecked size casts to uint8 here, and it's inconsistent why you use some uint32 for offset and some uint8 for sizes. I feel like the overall motivation here was that for some of these we know that they are constant small values and can fit in uint8. It's all fine because they do in fact fit, but it's odd to me that you check sometimes but not others.